sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala

```diff
@@ -120,6 +127,20 @@ class QueryPlanningTracker {
     ret
   }
 
+  /**
+   * Set when the query has been analysed and is ready for execution.
+   * This is after analysis for eager commands and after planning
+   * for other queries.
+   * see @link org.apache.spark.sql.execution.CommandExecutionMode
+   */
+  def setReadyForExecution(analyzedPlan: LogicalPlan): Unit = {
+    if (readyForExecution) {
+      throw new IllegalStateException("Cannot setReadyForExecution more than once")
+    }
```

Comment:
It passes tests, but I'm not 100% sure it would always hold. I think it's better to just `return` here to make sure that the callback is only called once. If you look at `QueryExecution.eagerlyExecutedCommands` it could have multiple commands in the query tree, that you e.g. union the results together or sth. Although I don't know of such use cases in practice, I think it's just OK to return here and make it be called the first time you reach executions.